### PR TITLE
yacc tool updates

### LINKED ISSUE #4326: GNU Bison options `--header` and `--defines` are not recognised by SCons when they don't have an argument
SCons ignores options `--header` and `--defines` passed to Bison when the argument is not provided.
The argument is optional for those options and if it is not provided SCons should behave the same way as if it encountered the options `-d` (which is detected correctly).
(https://www.gnu.org/software/bison/manual/bison.html#Output-Files)

* Link to SCons Users thread discussing your issue. - https://discord.com/channels/571796279483564041/571796280146133047/1087434623476301836
* Version of SCons - v4.5.1.49578b34582d9e92dac7d713a8e58599ae35aa63
* Version of Python - 3.10.6
* Which python distribution if applicable (python.org, cygwin, anaconda, macports, brew,etc) - the one from Ubuntu repository
* How you installed SCons - through `pip`
* What Platform are you on? (Linux/Windows and which version) - Linux Mint
* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues. - it's below
* How you invoke scons (The command line you're using "scons --flags some_arguments") - `scons`

```
targets = CFile('foo.y', YACCFLAGS='-d')
print('Flag `-d` (correct target list):', targets)

targets = CFile('bar.y', YACCFLAGS='--defines')
print('Flag `--defines` (incorrect target list):', targets)

targets = CFile('baz.y', YACCFLAGS='--header')
print('Flag `--header` (incorrect target list):', targets)
```

### LINKED ISSUE #4327: Add support for the GNU Bison option `-H`
SCons doesn't recognize the option `-H`.
It should update the list of output files when the option is encountered. The option `-H` is the same as `--header` or `--defines` which are currently supported. (Well, kinda. See: https://github.com/SCons/scons/issues/4326)
(https://www.gnu.org/software/bison/manual/bison.html#Output-Files)

* Link to SCons Users thread discussing your issue. - https://discord.com/channels/571796279483564041/571796280146133047/1087434623476301836
* Version of SCons - v4.5.1.49578b34582d9e92dac7d713a8e58599ae35aa63
* Version of Python - 3.10.6
* Which python distribution if applicable (python.org, cygwin, anaconda, macports, brew,etc) - the one from Ubuntu repository
* How you installed SCons - through `pip`
* What Platform are you on? (Linux/Windows and which version) - Linux Mint
* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues. - it's below
* How you invoke scons (The command line you're using "scons --flags some_arguments") - `scons`

```
targets = CFile('foo.y', YACCFLAGS='-d')
print('Flag `-d` (correct target list):', targets)

targets = CFile('bar.y', YACCFLAGS='-H')
print('Flag `-H` (incorrect target list):', targets)
```

### PR DESCRIPTION
**Pull request by mwichmann**
The yacc tool now recognizes the bison syntax of using `--header`, `--defines` and `--graph` options in `YACCFLAGS` without supplying an option argument. Note that byacc doesn't allow leaving off the option argument, and would error out in such cases.

Fixes #4326

The yacc tool now recognizes `-H` as a partial synonym for `--header`. It only works in the no-option-argument form - giving an option argument will likely break things.

Fixes #4327

Tests are expanded to cover these.

The graph file name (`-g`) is now generated relative to the requested target file name, not to the source name, to match actual current bison behavior. This is rare case - only kicks in if target explicitly requested with a different base name than source - in this case SCons would emit a different name than bison produces.  Unlikely to affect any real usage - you would have to set up the condition described, and additionally use the graph file as a source for some other build target.

The default file suffix for graph files is set to `.gv`, which is current for Bison as of 1.38.  The previous suffix for Bison, dating back to 2.4, was `.dot`, which is still the suffix byacc uses.  Docs updated to better describe when you need to set this. The former default, `.vcg`, has not been used since 2006.

Obsoleted `YACCVCGFILESUFFIX`, being replaced by `YACC_GRAPH_FILE_SUFFIX`.  If `YACC_GRAPH_FILE_SUFFIX` is not set, it will respect `YACCVCGFILESUFFIX`.

Documentation updated for the two header suffix construction variables to better describe (I hope) what these mean and when to set them. Other doc updates as well.

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -9,9 +9,16 @@ NOTE: 4.3.0 now requires Python 3.6.0 and above. Python 3.5.x is no longer suppo
 
 RELEASE  VERSION/DATE TO BE FILLED IN LATER
 
-      From John Doe:
-
-        - Whatever John Doe did.
+  From Mats Wichmann
```
Comment (bdbaddog) [inline CHANGES.txt:19]: add this blurb to RELEASE.txt please.

File: SCons/Tool/Tool.xml [hunk 2]
```diff
@@ -37,10 +37,19 @@ Example:
...
 # builds foo.c
-env.CFile(target = 'foo.c', source = 'foo.l')
+env.CFile(target='foo.c', source='foo.l')
+
 # builds bar.c
-env.CFile(target = 'bar', source = 'bar.y')
+env.CFile(target='bar', source='bar.y')
 </example_commands>
+
+<para>
```
Comment (bdbaddog) [inline SCons/Tool/Tool.xml]: Not sure we really need this notice.

File: SCons/Tool/Tool.xml [hunk 3]
```diff
@@ -59,10 +68,19 @@ Example:
...
-env.CXXFile(target = 'foo.cc', source = 'foo.ll')
+env.CXXFile(target='foo.cc', source='foo.ll')
+
 # builds bar.cc
-env.CXXFile(target = 'bar', source = 'bar.yy')
+env.CXXFile(target='bar', source='bar.yy')
 </example_commands>
+
+<para>
+Note that for yacc files,
```
Comment (bdbaddog) [inline SCons/Tool/Tool.xml:78]: Not sure we need this notice.

File: SCons/Tool/Tool.xml (hunk 2 above)
Comment (mwichmann) [reply SCons/Tool/Tool.xml]: Maybe not, but it could be a little surprising to traditional yacc users. Can certainly remove.

**Comment by bdbaddog** [comment]
can we add that blurb about files with the yacc tool, instaed of CFile() ?

**Comment by bdbaddog** [comment]
If a user is using an old version of bison or byacc would SCons' expected output files now be incorrect?

File: SCons/Tool/Tool.xml (hunk 2 above)
Comment (mwichmann) [reply SCons/Tool/Tool.xml]: Made the note non-visible.

**Comment by mwichmann** [comment]
> If a user is using an old version of bison or byacc would SCons' expected output files now be incorrect?

The only confusion could be about the graph file, where bison has changed file format once, and names twice; byacc has used the same format/suffix since they introduced the feature.  This only matters in the quite unlikely case that the graph file is used as the source to some other build - it's intended as a diagnostic output, not as something you'd use as part of a build. It only matters to us so there's something to "clean".  A bison old enough to notice the change would be 17+ years old.

**Comment by mwichmann** [comment]
I've updated (RELEASE, doc). Rebasing to current meant I had to force-push so things here will be out of sync.

**Comment by mwichmann** [comment]
The failure here is very odd:

https://ci.appveyor.com/project/SCons/scons/builds/46977299/job/dmkqe6n3umy68di0

The other windows runs didn't take this problem, and because it's an emulated tool (`myyacc.py`), not trying to use the real bison, it can't be a "tool is too old" problem. First guess is I might have used something that doesn't work in Python 3.6, but will have to investigate.

**Comment by jcbrill** [comment]
> First guess is I might have used something that doesn't work in Python 3.6, but will have to investigate.

After a REALLY QUICK (and possibly misguided) experiment, it appears that 3.6 needs a space after the "-I".

This fails in 3.6: "-Iout -Iin"
This passes in 3.6: "-I out -I in"

Both seem to work in 3.8.

**Comment by mwichmann** [comment]
Thanks for that. Maybe this is something in the `argparse` module - the mock tool didn't use that previously, it used `getopt` (I already forgot that change was made here, I thought I did that the last time I fiddled with this tool).

**Comment by bdbaddog** [comment]
> > If a user is using an old version of bison or byacc would SCons' expected output files now be incorrect?
> 
> The only confusion could be about the graph file, where bison has changed file format once, and names twice; byacc has used the same format/suffix since they introduced the feature. This only matters in the quite unlikely case that the graph file is used as the source to some other build - it's intended as a diagnostic output, not as something you'd use as part of a build. It only matters to us so there's something to "clean". A bison old enough to notice the change would be 17+ years old.

Couldn't this be used to generate some document for the software being built?
To generate the correct output file, we'd need to interrogate the tool to see which version is being used?

**Comment by mwichmann** [comment]
Maybe wait for someone to request that?

**Comment by mwichmann** [comment]
To follow up on the error a bit, here's the thing that goes wrong extracted - I can reproduce this fail on Linux also:
```py
$ python3.6 myyacc.py -x -Iout -Iin -o out/aaa.c in/aaa.y
usage: myyacc.py [-h] [-g [GRAPH]] [-d] [-v] [-x] [-H [HEADER]] -o OUT
                 [-I I_ARGUMENTS]
                 files [files ...]
myyacc.py: error: unrecognized arguments: -Iout -Iin
$ python3.7 myyacc.py -x -Iout -Iin -o out/aaa.c in/aaa.y
usage: myyacc.py [-h] [-g [GRAPH]] [-d] [-v] [-x] [-H [HEADER]] -o OUT
                 [-I I_ARGUMENTS]
                 files [files ...]
myyacc.py: error: unrecognized arguments: -Iout -Iin
$ python3.8 myyacc.py -x -Iout -Iin -o out/aaa.c in/aaa.y
$
```
So 3.6 and 3.7 fail this, 3.8 and later passes. The argparse docs don't list any 3.8 behavioral change that could affect this, so presumably it was a "bugfix".

Since the argument isn't a yacc argument but something specific to the mock command, we can fix this up in the test by including the space so it works on all versions. The mock command needs a tweak for that.

**Comment by bdbaddog** [comment]
So here's my concern. Often parsers depend on a specific version of bison/yacc/byacc. Changing the version may introduce changes which will break the logic.
Bison 3.8 is a 2021 release, so relatively recent.

But as you said this is only for the graph output file?
So maybe just a more explicit note
NOTE: If you are using Bison < 3.8, and outputting a graph file, SCons will expect the incorrect version. Please send email to users@scons.org or let us know on our discord server.

**Comment by mwichmann** [comment]
I suppose.  bison switched way long ago to using `.dot`. byacc then followed suit.  Then the graphviz folks told the bison folks "don't do that, `.dot` is ambiguous as a suffix".  byacc didn't follow suit.  Leaves us with a fairly impossible situation.

The `.gv` suffix has been around quite a bit longer, this just flipped the suffix default "if you say nothing".
```
+  In conformance with the recommendations of the GraphViz team, if %require
+  "3.4" (or better) is specified, the option --graph generates a *.gv file
+  by default, instead of *.dot.
```
Further, you can specify the filename you want via other mechanisms. These are just ugly compat things that don't really work well with SCons anyway - `$YACC_GRAPH_FILE` is preferred.

Does that make you any happier?

**Comment by mwichmann** [comment]
Cleaned up conflicts.

**Pull request merged**